system/settings/settings.c

```diff
@@ -0,0 +1,1432 @@
+
+  setting->type = SETTING_IP_ADDR;
+  memcpy(&setting->ip, ip, sizeof(struct in_addr));
+
+  return OK;
+}
+
+/****************************************************************************
+ * Name: load
+ *
+ * Description:
+ *    Loads all values
+ *
+ * Input Parameters:
+ *    none
+ *
+ * Returned Value:
+ *   Success or negated failure code
+ *
+ ****************************************************************************/
+
+static int load(void)
+{
+  int i;
+  int ret = OK;
+
+  for (i = 0; i < CONFIG_SETTINGS_MAX_STORAGES; i++)
+    {
+      if ((g_settings.store[i].file[0] != '\0') &&
+           g_settings.store[i].load_fn)
+        {
+          ret = g_settings.store[i].load_fn(g_settings.store[i].file);
+          if (ret < 0)
+            {
+              break;
+            }
+        }
+    }
+
+  return ret;
+}
+
+/****************************************************************************
+ * Name: save
+ *
+ * Description:
+ *    Saves cached values, either immediately or on a timer
+ *
+ * Input Parameters:
+ *    none
+ *
+ * Returned Value:
+ *   Success or negated failure code
+ *
+ ****************************************************************************/
+
+static int save(void)
+{
+  int ret = OK;
+  g_settings.wrpend = 1;
+
+#ifdef CONFIG_SETTINGS_CACHED_SAVES
+  timer_settime(g_settings.timerid, 0, &g_settings.trigger, NULL);
+#else
+  union sigval value =
+  {
+    .sival_int = 0,
+  };
+
+  ret = dump_cache(value);
+#endif
+
+  return ret;
+}
+
+/****************************************************************************
+ * Name: signotify
+ *
+ * Description:
+ *    Notify anything waiting for a signal
+ *
+ * Input Parameters:
+ *    none
+ *
+ * Returned Value:
+ *    none
+ *
+ ****************************************************************************/
+
+static void signotify(void)
+{
+  int i;
+
+  for (i = 0; i < CONFIG_SETTINGS_MAX_SIGNALS; i++)
+    {
+      if (g_settings.notify[i].pid == 0)
+        {
+          break;
+        }
+
+      kill(g_settings.notify[i].pid, g_settings.notify[i].signo);
+    }
+}
+
+/****************************************************************************
+ * Name: dump_cache
+ *
+ * Description:
+ *    Writes out the cached data to the appropriate storage
+ *
+ * Input Parameters:
+ *    value            - parameter passed if called from a timer signal
+ *
+ * Returned Value:
+ *   Success or negated failure code
+ *
+ ****************************************************************************/
+
+static int dump_cache(union sigval value)
+{
+  int ret = OK;
+  (void)value;
+  int i;
+
+  ret = pthread_mutex_lock(&g_settings.mtx);
+  if (ret < 0)
+    {
+      return ret;
+    }
+
+  for (i = 0; i < CONFIG_SETTINGS_MAX_STORAGES; i++)
+    {
+      if ((g_settings.store[i].file[0] != '\0') &&
+           g_settings.store[i].save_fn)
+        {
+          ret = g_settings.store[i].save_fn(g_settings.store[i].file);
```

Comment:
it isn't safe to call fs function in signal handler

Reply:
@xiaoxiang781216 - OK - will look tomorrow as need a clear mind for it :)

Reply:
@xiaoxiang781216 - can you give me a little more information on what's not safe, so I can devise the best fix, please?

Reply:
here: https://man7.org/linux/man-pages/man7/signal-safety.7.html.

Reply:
Found that as soon as I replied a few minutes ago so deleted it. Thanks :)

Reply:
@xiaoxiang781216 If I have understood this right:

1. Calls to the save function are obviously asynchronous
2. Each call to the save function sets a a new timer
3. That timer could trigger before the processing of a previous timer has completed. And that potentially violates the signal safety concern?

I have proven this to be an issue, to satisfy my curiosity by adding a sleep into the the save function and calling it a few times before the function awakens. It crashes, so - I hope - I've understood it?

@fjpanag Thoughts?

Reply:
@fjpanag @xiaoxiang781216 
Maybe the save() function should  only set a new timer if g_settings.wrpend == 0?

I am obviously struggling a little here, sorry :(

Reply:
Hmmm...been reading more....the save and/or dump functions need to become reentrant. Will take another look tomorrow, but still open to suggestions :)

Reply:
@fjpanag @xiaoxiang781216 
The more I think about this the more I wondering whether the "dump_cache" function should be removed completely?

The code as-is presumes a write to a block device (e.g. SD card) but that may not be what's wanted. In my case, I actually need to write the data to EEPROM, and only bytes that have changed, so that needs a separate and unrelated thread to handle it.

This code currently has the ability to signal other threads that c hange has occured. Is that not enough? Off load any async file saving away from this settings function?

I am also inclined to move away from a signal and move to a poll method instead?

What do you both think?

Otherwise, I think I should change it back to a draft PR -  or even close it completely and forgot it? If it's not doing things "right" then I don't want to use it, nor will others, and it certainly isn't going to be merged of course.